packages/aws-cdk/lib/api/hotswap/stepfunctions-state-machines.ts

```diff
@@ -0,0 +1,72 @@
+import { ISDK } from '../aws-auth';
+import { ChangeHotswapImpact, ChangeHotswapResult, HotswapOperation, HotswappableResourceChange, establishHotswappableResourceName } from './common';
+import { EvaluateCloudFormationTemplate } from './evaluate-cloudformation-template';
+
+export async function isHotswappableStateMachineChange(
+  logicalId: string, change: HotswappableResourceChange, evaluateCfnTemplate: EvaluateCloudFormationTemplate,
+): Promise<ChangeHotswapResult> {
+  const stateMachineDefinitionChange = await isStateMachineDefinitionOnlyChange(change, evaluateCfnTemplate);
+
+  if ((stateMachineDefinitionChange === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) ||
+      (stateMachineDefinitionChange === ChangeHotswapImpact.IRRELEVANT)) {
+    return stateMachineDefinitionChange;
+  }
+
+  const machineNameInCfnTemplate = change.newValue?.Properties?.StateMachineName;
+  const machineName = await establishHotswappableResourceName(logicalId, machineNameInCfnTemplate, evaluateCfnTemplate);
+
+  if (!machineName) {
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+
+  return new StateMachineHotswapOperation({
+    definition: stateMachineDefinitionChange,
+    stateMachineName: machineName,
+  });
+}
+
+async function isStateMachineDefinitionOnlyChange(
+  change: HotswappableResourceChange, evaluateCfnTemplate: EvaluateCloudFormationTemplate):
+  Promise<string | ChangeHotswapImpact> {
+  const newResourceType = change.newValue.Type;
+  if (newResourceType !== 'AWS::StepFunctions::StateMachine') {
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+
+  const propertyUpdates = change.propertyUpdates;
+
+  for (const updatedPropName in propertyUpdates) {
+    const updatedProp = propertyUpdates[updatedPropName];
+    if (updatedProp.newValue === undefined) {
+      return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+    }
```

Comment:
What's the purpose of this check? I feel like we can safely get rid of this one. Or at least move it after the `'DefinitionString'` check - not sure if there's a case for updating your State Machine with an empty `definitionString`.